File: src/Shared/SpeechBalloonView.m
//
//  SpeechBalloonView.m
//  OpenStreetMap
//
//  Created by Bryce Cogswell on 12/11/12.
//  Copyright (c) 2012 Bryce Cogswell. All rights reserved.
//

#import "SpeechBalloonLayer.h"
#import "SpeechBalloonView.h"

static const CGFloat arrowWidth = 16;
static const CGFloat arrowHeight = 16;

@implementation SpeechBalloonView

+ (Class)layerClass
{
	return [CAShapeLayer class];
}

- (id)initWithText:(NSString *)text balloonPress:(void(^)(void))balloonPress disclosurePress:(void (^)(void))disclosurePress
{
	self = [super initWithFrame:CGRectMake(0,0,0,0)];
	if ( self ) {

		_balloonPress = balloonPress;
		_disclosurePress = disclosurePress;

#if !TARGET_OS_IPHONE
		self.wantsLayer = YES;
#endif
		CAShapeLayer * shapeLayer = (id)self.layer;

		// shape layer
		shapeLayer.fillColor = NSColor.grayColor.CGColor;
		shapeLayer.strokeColor = NSColor.blackColor.CGColor;

		// disclosure button view
#if TARGET_OS_IPHONE
		UIButton * button = [UIButton buttonWithType:UIButtonTypeDetailDisclosure];
		CGRect buttonRect = button.frame;
		[button addTarget:self action:@selector(disclosurePress:) forControlEvents:UIControlEventTouchUpInside];
#else
		CGRect buttonRect = CGRectMake( 0, 0, 20, 20);
		NSButton * button = [[NSButton alloc] initWithFrame:buttonRect];
		[button setButtonType:NSOnOffButton];
		[button setBezelStyle:NSDisclosureBezelStyle];
		[button highlight:NO];
		[button setTitle:nil];
		button.wantsLayer = YES;
#endif

		// text layer
		CATextLayer * textLayer = [CATextLayer layer];
#if TARGET_OS_IPHONE
		CGFontRef font = CGFontCreateWithFontName( (__bridge CFStringRef) @"Helvetica-Bold" );
		textLayer.font = font;
		CGFontRelease(font);
#else
		NSFont * font = [[NSFontManager sharedFontManager] convertFont:[NSFont labelFontOfSize:12] toHaveTrait:NSBoldFontMask];
		textLayer.font = (__bridge CFTypeRef)font;
#endif
		textLayer.fontSize = 16;
		textLayer.alignmentMode = kCAAlignmentLeft;
		textLayer.string = text;
		textLayer.foregroundColor = NSColor.whiteColor.CGColor;
		[shapeLayer addSublayer:textLayer];

		CGSize textSize = textLayer.preferredFrameSize;

		CGSize boxSize = textSize;
		boxSize.width += 35;
		boxSize.height = 30;

		// creat path with arrow
		const CGFloat cornerRadius = 4;
		_path = CGPathCreateMutable();
		CGPathMoveToPoint(_path, NULL, boxSize.width/2, boxSize.height+arrowHeight);	// arrow bottom
		CGPathAddLineToPoint(_path, NULL, boxSize.width/2-arrowWidth/2, boxSize.height);	// arrow top-left
		CGPathAddArcToPoint(_path, NULL, 0, boxSize.height, 0, 0, cornerRadius);	// bottom right corner
		CGPathAddArcToPoint(_path, NULL, 0, 0, boxSize.width, 0, cornerRadius);	// top left corner
		CGPathAddArcToPoint(_path, NULL, boxSize.width, 0, boxSize.width, boxSize.height, cornerRadius); // top right corner
		CGPathAddArcToPoint(_path, NULL, boxSize.width, boxSize.height, 0, boxSize.height, cornerRadius);	// bottom right corner
		CGPathAddLineToPoint(_path, NULL, boxSize.width/2+arrowWidth/2, boxSize.height );	// arrow top-right
		CGPathCloseSubpath(_path);
		CGRect viewRect = CGPathGetPathBoundingBox( _path );
		shapeLayer.path = _path;

		buttonRect = CGRectOffset( buttonRect, boxSize.width - buttonRect.size.width, 0 );
		button.frame = buttonRect;
		[self addSubview:button];

		textLayer.frame = CGRectMake(5, 5, boxSize.width - 5, boxSize.height - 5);

		self.frame = CGRectMake(0, 0, viewRect.size.width, viewRect.size.height);
	}
	return self;
}

- (void)disclosurePress:(id)sender
{
	_disclosurePress();
}

- (void) setPoint:(CGPoint)point
{
	// set bottom center at point
	CGRect rect = self.frame;
	rect.origin.x = point.x - rect.size.width / 2;
	rect.origin.y = point.y - rect.size.height - 3;
	self.frame = rect;
}


- (BOOL)pointInside:(CGPoint)point withEvent:(NSEvent *)event
{
#if TARGET_OS_IPHONE
	if ( ![super pointInside:point withEvent:event] )
		return NO;
	if ( ! CGPathContainsPoint( _path, NULL, point, NO) )
		return NO;
	return YES;
#else
	return NO;
#endif
}



- (void)dealloc
{
	CGPathRelease( _path );
}

@end

File: src/Shared/SpeechBalloonView.swift
//
//  SpeechBalloonView.swift
//  OpenStreetMap
//
//  Created by Bryce Cogswell on 12/11/12.
//  Copyright (c) 2012 Bryce Cogswell. All rights reserved.
//

import Foundation
import UIKit

class SpeechBalloonView: UIView {
	let path: CGPath
	let arrowWidth: CGFloat = 20
	let arrowHeight: CGFloat = 48

	class func layerClass() -> AnyClass {
		return CAShapeLayer.self
	}

	@available(*, unavailable)
	required init?(coder: NSCoder) {
		fatalError("init(coder:) has not been implemented")
	}

	init(text: String?) {
		// text layer
		let textLayer = CATextLayer()
		textLayer.contentsScale = UIScreen.main.scale
		let font = UIFont.preferredFont(forTextStyle: .headline)
		textLayer.font = font
		textLayer.fontSize = 18
		textLayer.alignmentMode = .center
		textLayer.string = text
		textLayer.foregroundColor = UIColor.black.cgColor

		let textSize = textLayer.preferredFrameSize()

		var boxSize = textSize
		boxSize.width += 35
		boxSize.height += 30

		// creat path with arrow
		let cornerRadius: CGFloat = 14
		let center = 0.35
		let path = CGMutablePath()
		path.move(to: CGPoint(x: boxSize.width / 2, y: boxSize.height + arrowHeight)) // arrow bottom
		path
			.addLine(to: CGPoint(x: CGFloat(Double(boxSize.width) * center - Double(arrowWidth / 2)),
			                     y: boxSize.height)) // arrow top-left
		path.addArc(
			tangent1End: CGPoint(x: 0, y: boxSize.height),
			tangent2End: CGPoint(x: 0, y: 0),
			radius: cornerRadius) // bottom right corner
		path.addArc(
			tangent1End: CGPoint(x: 0, y: 0),
			tangent2End: CGPoint(x: boxSize.width, y: 0),
			radius: cornerRadius) // top left corner
		path.addArc(
			tangent1End: CGPoint(x: boxSize.width, y: 0),
			tangent2End: CGPoint(x: boxSize.width, y: boxSize.height),
			radius: cornerRadius) // top right corner
		path.addArc(
			tangent1End: CGPoint(x: boxSize.width, y: boxSize.height),
			tangent2End: CGPoint(x: 0, y: boxSize.height),
			radius: cornerRadius) // bottom right corner
		path
			.addLine(to: CGPoint(x: CGFloat(Double(boxSize.width) * center + Double(arrowWidth / 2)),
			                     y: boxSize.height)) // arrow top-right
		path.closeSubpath()
		self.path = path

		textLayer.frame = CGRect(
			x: (boxSize.width - textSize.width) / 2,
			y: (boxSize.height - textSize.height) / 2,
			width: textSize.width,
			height: textSize.height)

		let viewRect = path.boundingBoxOfPath
		super.init(frame: CGRect(x: 0, y: 0, width: 0, height: 0))
		frame = CGRect(x: 0, y: 0, width: viewRect.size.width, height: viewRect.size.height)

		let shapeLayer = layer as! CAShapeLayer
		shapeLayer.path = path

		// shape layer
		shapeLayer.fillColor = UIColor.white.cgColor
		shapeLayer.strokeColor = UIColor.black.cgColor
		shapeLayer.lineWidth = 6

		shapeLayer.addSublayer(textLayer)
	}

	func setPoint(_ point: CGPoint) {
		// set bottom center at point
		var rect = frame
		rect.origin.x = point.x - rect.size.width / 2
		rect.origin.y = point.y - rect.size.height
		frame = rect
	}

	func setTarget(_ view: UIView?) {
		let rc = view?.frame
		let pt = CGPoint(
			x: Double((rc?.origin.x ?? 0.0) + (rc?.size.width ?? 0.0) / 2),
			y: Double((rc?.origin.y ?? 0.0) - (rc?.size.height ?? 0.0) / 2))
		setPoint(pt)
	}

	override func point(inside point: CGPoint, with event: UIEvent?) -> Bool {
		if !super.point(inside: point, with: event) {
			return false
		}
		return path.contains(point)
	}
}
